llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp

```diff
@@ -0,0 +1,412 @@
+    RISCVBenchmarkOptions("llvm-exegesis benchmark RISCV options");
+
+// TODO move perf counter data to td files (although it looks like an overkill
+// of sorts)
+
+static const char *RISCVPfmCounterNames[] = {
+    "CPU_CYCLES", // 0
+};
+
+static const PfmCountersInfo RISCVDefaultPfmCounters = {
+    RISCVPfmCounterNames[0], // Cycle counter
+    nullptr,                 // No uops counter.
+    nullptr,                 // No issue counters.
+    0};
+
+static const CpuAndPfmCounters RISCVCpuPfmCounters[] = {
+    {"", &RISCVDefaultPfmCounters},
+};
+
+class ExegesisRISCVTarget : public ExegesisTarget {
+public:
+  ExegesisRISCVTarget();
+
+  Expected<std::unique_ptr<pfm::CounterGroup>>
+  createCounter(StringRef CounterName, const LLVMState &State,
+                ArrayRef<const char *> ValidationCounters,
+                const pid_t ProcessID) const override;
+
+  bool checkOpcodeSupported(int Opcode,
+                            const MCSubtargetInfo &SI) const override;
+
+  unsigned findRegisterByName(const StringRef RegName) const override;
+
+  bool matchesArch(Triple::ArchType Arch) const override;
+
+  std::vector<MCInst> setRegTo(const MCSubtargetInfo &STI, unsigned Reg,
+                               const APInt &Value) const override;
+
+  unsigned getDefaultLoopCounterRegister(const Triple &) const override;
+
+  void decrementLoopCounterAndJump(MachineBasicBlock &MBB,
+                                   MachineBasicBlock &TargetMBB,
+                                   const MCInstrInfo &MII,
+                                   unsigned LoopRegister) const override;
+
+  unsigned getScratchMemoryRegister(const Triple &TT) const override;
+
+  void fillMemoryOperands(InstructionTemplate &IT, unsigned Reg,
+                          unsigned Offset) const override;
+
+  virtual std::vector<MCInst>
+  storeRegValueToScratch(const MCSubtargetInfo &STI, unsigned Reg,
+                         unsigned Offset) const override;
+  ArrayRef<unsigned> getUnavailableRegisters() const override;
+
+  Error randomizeTargetMCOperand(const Instruction &Instr, const Variable &Var,
+                                 MCOperand &AssignedValue,
+                                 const BitVector &ForbiddenRegs) const override;
+
+  void processInstructionReservedRegs(InstructionTemplate &IT) const override;
+
+  std::vector<InstructionTemplate>
+  generateInstructionVariants(const Instruction &Instr,
+                              unsigned MaxConfigsPerOpcode) const override;
+};
+
+ExegesisRISCVTarget::ExegesisRISCVTarget()
+    : ExegesisTarget(RISCVCpuPfmCounters, RISCV_MC::isOpcodeAvailable) {}
+
+Expected<std::unique_ptr<pfm::CounterGroup>> ExegesisRISCVTarget::createCounter(
+    StringRef CounterName, const LLVMState &State,
+    ArrayRef<const char *> ValidationCounters, const pid_t ProcessID) const {
+  if (CounterName == RISCVPfmCounterNames[0]) {
+    return createRISCVCpuCyclesCounter(pfm::PerfEvent(CounterName));
```

Comment:
I'm not too familiar with perf support on RISCV (so please correct me if I'm mistaken here), but it seems like some things (like a cycles counter) might be supported? It might be reasonable to allow for using the perf subsystem in addition to the custom counter similar to how X86 supports the perf subsystem and LBR.

Reply:
I also think we should try to use as much Linux Perf as possible. Granted, reading `mcycle` from CSR register is more generic, but llvm-exegesis already puts an assumption that it has to run on Linux so we might as well just use Perf. Libpfm hasn't supported RISC-V yet but we can use Perf to read the CYCLES event.

Reply:
We'd still have to go through libpfm to get the event ID/description to do a call to `perf_event_open` with how things are structured currently, but there should be generic Linux perf events that can be used (see [here](https://github.com/wcohen/libpfm4/blob/1c0cdb91bc79eb1bc827e022f0a3738f124796a2/lib/events/perf_events.h)). Probably will want to run [this example](https://github.com/wcohen/libpfm4/blob/1c0cdb91bc79eb1bc827e022f0a3738f124796a2/examples/showevtinfo.c) though to see what events are there as it isn't super clear from the header files.

Reply:
Thanks for the offer. We will definitely look into this, but I suggest putting it in a separate pool request. Since it does not violate the functionality of exegesis, and the changes are significant.